jellyfish/_notifier.py

```diff
@@ -0,0 +1,130 @@
+        self._websocket = None
+        self._ready = False
+
+        self._ready_event: asyncio.Event = None
+
+        self._notification_handler: Callable = None
+        self._metrics_handler: Callable = None
+
+    def on_server_notification(self, handler: Callable[[Any], None]):
+        '''
+        Decorator used for defining handler for ServerNotifications
+        i.e. all messages other than `ServerMessageMetricsReport`.
+        '''
+        self._notification_handler = handler
+        return handler
+
+    def on_metrics(self, handler: Callable[[ServerMessageMetricsReport], None]):
+        '''
+        Decorator used for defining handler for `ServerMessageMetricsReport`.
+        '''
+        self._metrics_handler = handler
+        return handler
+
+    async def connect(self):
+        '''
+        A coroutine which connects Notifier to Jellyfish and listens for all incoming
+        messages from the Jellyfish.
+
+        It runs until the connection isn't closed.
+
+        The incoming messages are handled by the functions defined using the
+        `on_server_notification` and `on_metrics` decorators.
+        '''
+        async with client.connect(f'ws://{self._server_address}/socket/server/websocket') \
+                as websocket:
+            try:
+                self._websocket = websocket
+                await self._authenticate()
+
+                if self._notification_handler:
+                    await self._subscribe_event(
+                        event=ServerMessageEventType.EVENT_TYPE_SERVER_NOTIFICATION)
+
+                if self._metrics_handler:
+                    await self._subscribe_event(event=ServerMessageEventType.EVENT_TYPE_METRICS)
+
+                self._ready = True
+                if self._ready_event:
+                    self._ready_event.set()
+
+                await self._receive_loop()
+            finally:
+                self._websocket = None
+
+    async def wait_ready(self) -> True:
+        '''
+        Waits until the notifier is connected and authenticated to Jellyfish.
+
+        If already connected, returns `True` immediately.
+        '''
+        if self._ready:
+            return True
+
+        if self._ready_event is None:
+            self._ready_event = asyncio.Event()
+
+        await self._ready_event.wait()
+
+    async def _authenticate(self):
+        msg = ServerMessage(auth_request=ServerMessageAuthRequest(token=self._server_api_token))
+        await self._websocket.send(bytes(msg))
+
+        try:
+            message = await self._websocket.recv()
+        except ConnectionClosed as exception:
+            if 'invalid token' in str(exception):
+                raise RuntimeError('Invalid server_api_token') from exception
+            raise
+
+        message = ServerMessage().parse(message)
+
+        _type, message = betterproto.which_one_of(message, 'content')
+        assert isinstance(message, ServerMessageAuthenticated)
```

Comment:
I know next to nothing about 'production-grade' Python, but should we really use `assert`s as sanity checks?

Reply:
according to this discussion: https://stackoverflow.com/questions/944592/best-practice-for-using-assert
`assert` should be used to _check for things that should never happen_ .

I think this is an appropriate use case, since if this doesn't match then we have some major change in Jellyfish.